### PR TITLE
Post/flow

### PR DESCRIPTION
**Pull request by dvdzkwsk**

### CODE REVIEW

**Comment by dvdzkwsk** [comment]
Happy birthday @Fluidbyte

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 1]
```diff
@@ -0,0 +1,372 @@
+---
+layout: post
+title: "Avoid Runtime Errors in React with Flow"
+date: 2016-02-29 18:00:00
+categories: reactjs flow javascript development
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: Later we may want to update the taxonomy to leverage tags in addition to categories so we can keep cats high-level (e.g. development, design, business intelligence) while allowing tags to be used for keyword relevance and a [posts by tag page](http://habd.as/tags/), for example.

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+author: david.zukowski
+cover: /assets/images/covers/water.jpg
+description: Catch errors before they happen by statically analyzing your code with Flow.
+---
+
+[Flow](http://flowtype.org/) is pretty awesome.
+
+Oh, more details... Ok, so, Flow is great for large or growing codebases that could use a hand in the [static analysis](https://en.wikipedia.org/wiki/Static_program_analysis) department. Even outside of the context of React ([heh](https://facebook.github.io/react/docs/context.html)), Flow shines by ensuring type safety across modules; that is, it allows you to annotate variables, functions, et. al. with type annotations that Flow can parse and use to spot errors in your application before they happen.
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: Small typo with `et al.`

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+Why Flow? Our team settled on Flow because we already had multiple fairly substantial React codebases with complicated-enough build systems that we wanted to avoid modifying (the build systems), which pretty much discounted TypeScript, at least for the immediate gains we were seeking. Luckily, Flow gave us the opportunity to incrementally add types to our applications with a simple `/* @flow */` pragma and little else.
+
+If you haven't given it a chance yet I strongly advise [giving it five minutes](https://signalvnoise.com/posts/3124-give-it-five-minutes). In fact, despite balking at some of the syntax at first, the first time I saw a `Property cannot be accessed on possibly null or undefined value` on a [Maybe Type](http://flowtype.org/docs/nullable-types.html#_) I pretty much spammed our team's Slack channel in excitement.
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: I love the give it 5 minutes post!

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+1) `PropTypes` are not statically analyzable. When you encounter an error it's during development and only once the problem component has rendered. That's far too late, and the fact that it's caught at runtime means you're likely to miss errors that could end up biting you when you roll out to production.
+
+2) They add unneeded `noop` bytes in production. Yes, not the biggest of deals, and it [is possible to remove them](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types), but that's Yet Another Plugin for your growing webpack + babel + rosetta stone build pipeline.
+
+3) As the name implies, the type checking is limited to component properties: they cannot be used to type check any other parts of your codebase.
+
+4) Finally, `PropTypes` are simply less powerful. They can, to a degree, ensure that the props entering a component match your expectations, but that's where their utility ends; they can make no guarantees that you, for example, do not try to access non-existent properties on objects passed through props, nor do `PropTypes` check _how_ those props are used throughout a component.
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: Should this be an unordered list of items?

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+It was, as the thesis states, awesome.
+
+## Type Safety and React
+
+Unlike what some [recent rumors](https://www.reddit.com/r/reactjs/comments/44yqge/fyi_proptypes_are_on_their_way_out/czubij1) may have implied, `PropTypes` are not going anywhere, at least any time soon. However, that doesn't mean they're perfect; there are a number of places where they fall short, notably:
+
+1) `PropTypes` are not statically analyzable. When you encounter an error it's during development and only once the problem component has rendered. That's far too late, and the fact that it's caught at runtime means you're likely to miss errors that could end up biting you when you roll out to production.
+
+2) They add unneeded `noop` bytes in production. Yes, not the biggest of deals, and it [is possible to remove them](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types), but that's Yet Another Plugin for your growing webpack + babel + rosetta stone build pipeline.
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: YAP? ^^

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+3) As the name implies, the type checking is limited to component properties: they cannot be used to type check any other parts of your codebase.
+
+4) Finally, `PropTypes` are simply less powerful. They can, to a degree, ensure that the props entering a component match your expectations, but that's where their utility ends; they can make no guarantees that you, for example, do not try to access non-existent properties on objects passed through props, nor do `PropTypes` check _how_ those props are used throughout a component.
+
+Flow solves all of these concerns and is an invaluable addition to the typical workflow, perhaps as a [pre-commit hook](http://technologyadvice.github.io/the-power-of-the-precommit/) or as part of your CI process. But enough with the words already, let's write some code. Everyone is rightfully sick of [TodoMVC](http://todomvc.com/), so we'll change it up by doing some TodoMVC _with Flow_.
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: Nice internal backlink! The SEO crowd is smiling. If desired, you can remove the `http://technologyadvice.github.io` from the beginning of the link to make the link root-relative.

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+type Props = {
+  todo: Object,
+  onComplete: (todo: Todo) => any,
+};
+{% endhighlight %}
+
+At this point, we've defined the property `onComplete` to be a function that takes a `Todo` as its first and only argument. We don't actually care what that function returns, it's simply a callback, and those guys are super forgettable, so we specify `any` as the return type. With this in place, let's run `flow check` and see what we get:
+
+<div style="text-align:center;margin: 1rem auto;">
+  <img src="http://i.imgur.com/wBsGhTM.png"/>
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: Preferably we would store all images inside this app to keep all of the content consolidated in one place.

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+      // "The Witch" is pretty good, you should see it.
+      someBadProperty: 'Black Phillip',
+    })
+  }
+  // ...
+}
+{% endhighlight %}
+
+The amazing part about this setup is that if `<MyFormComponent />` decides to not abide by the shape of the filters (as it fails to do above), Flow will catch it without ever having to run the application. Whether the error is an additional/malformed property, a missing required property, or an incorrect type, Flow has you covered. If you ever change the shape your state, Flow will help you confirm that the rest the application continues humming along.
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: You may want to reinforce your point here about this kind of error leaking into production without flow.

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+Flow isn't perfect, it doesn't catch everything just yet and sometimes requires you to write your code in a way that's a bit more explicit, but it's a worthy tradeoff, especially considering the fact that it's actively maintained and constantly improving.
+
+Seriously, Flow is awesome.
```
Comment (ghost) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: CTA?

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown (hunk 7 above)
Comment (ksafranski) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: Agreed, let's create a `posts` dir in the assets and if a post has assets create a dir by the same name in that dir.

**Comment by ghost** [comment]
Just a few small notes. Otherwise @davezuko is boss. Looking forward to promoting this post.

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown (hunk 2 above)
Comment (dvdzkwsk) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: fixed, thanks!

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown (hunk 4 above)
Comment (dvdzkwsk) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: fixed

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown (hunk 7 above)
Comment (dvdzkwsk) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: fixed

File: _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown (hunk 8 above)
Comment (dvdzkwsk) [inline _posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown]: :+1: will do

**Comment by ksafranski** [comment]
:accept:

**Pull request merged**